agent/api/task/task.go

```diff
@@ -3541,14 +3545,53 @@ func (task *Task) PopulateServiceConnectContainerMappingEnvVar() error {
 		if err != nil {
 			return fmt.Errorf("error retrieving task container for pause container %s: %+v", c.Name, err)
 		}
-		containerMapping[taskContainer.Name] = c.GetNetworkSettings().IPAddress
+		if instanceIPCompatibility.IsIPv6Only() && c.GetNetworkSettings().GlobalIPv6Address != "" {
+			containerMapping[taskContainer.Name] = c.GetNetworkSettings().GlobalIPv6Address
+		} else {
+			containerMapping[taskContainer.Name] = c.GetNetworkSettings().IPAddress
+		}
+	}
+	return task.setContainerMappingForServiceConnectContainer(containerMapping)
+}
+
+// PopulateServiceConnectContainerMappingEnvVarAwsvpc populates APPNET_CONTAINER_IP_MAPPING
+// environment variable for the AppNet Agent container (aka Service Connect container) for
+// awsvpc network mode tasks.
+// Application containers are expected to be bound to loopback address(es).
+func (task *Task) PopulateServiceConnectContainerMappingEnvVarAwsvpc() error {
+	primaryENI := task.GetPrimaryENI()
+	if primaryENI == nil {
+		return errors.New("no primary ENI found in task")
+	}
+	if !primaryENI.IsIPv6Only() {
```

Comment:
Can the sidecar container be smart enough to check IPv6 only environment, so that PopulateServiceConnectContainerMappingEnvVarAwsvpc may not be needed in Agent? I wonder why we would need to do it only for IPv6 and let the default case of IPv4 be handled by the sidecar container itself?

Reply:
It is not trivial for the sidecar container to figure out IP compatibility as it does not have task ENI information. Today there is no IP mapping for awsvpc tasks because it is assumed that the application containers are bound to IPv4 loopback and hence the sidecar container's default behavior. This mapping overrides that behavior.